Feat/cvc

chrismamo1 opened the pull request.

jcm-cryptosense committed:
added CVC parsing for RSA and ECDSA

jcm-cryptosense committed:
added tests and optimizations for CVC parsing

comments

jcm-cryptosense committed:
update changelog

jcm-cryptosense committed:
made CVC.decode return a Result.result value

jcm-cryptosense committed:
style

emillon commented on src/asn1.ml:
```diff
@@ -615,11 +667,263 @@ struct
 let default_result default = function Result.Error _ -> default () | Result.Ok _ as r -> r
 
+module CVC = struct
+  let rsa_oid = Asn.OID.of_string "0.4.0.127.0.7.2.2.2.1.4"
```
What do you think about naming the root (`0.4.0.127.0.7.2.2.2 id-TA Terminal Authentication`) and using `<|` to build these two OIDs?

chrismamo1 replied:
I'm not sure how much that would improve readability, and I'm a bit apprehensive to use custom operators.

emillon replied:
I'm fine with the `of_string` solution, but here are a couple benefits in doing otherwise:
- that documents the name of the root OID
- that removes a bit of duplication (not super important but typos happen)
- this changes from running a parser at runtime to using a constructor. That is the same difference as writing `Int64.of_string "12345"` vs writing `12345L`. With a local open, there's little danger in using custom operator (`Asn_oid.(id_ta_terminal_authentication <|| [1;4])` is almost an OID litteral).

emillon commented on src/asn1.ml:
```diff
@@ -615,11 +667,263 @@ struct
+      let code =
+        let msb = code * 0x100 in
+        let lsb =
+          Cstruct.get_char tag 1
```
`Cstruct.get_uint8` will give you an `int` without having to resort to `Char.code` (same above).

chrismamo1 replied:
Changed in all instances. Thanks!

emillon commented on src/asn1.ml:
```diff
@@ -615,11 +667,263 @@ struct
+    map f g oid
+
+  let decode_oid str =
+    let t, left = Asn.(decode_exn (codec ber grammar) str) in
```
You should check for leftover here (I think there's a helper somewhere to do that).

chrismamo1 replied:
I couldn't find a specific helper for reporting leftover errors, so I used the method established elsewhere in the library.

emillon replied:
OK, that might be worth extracting one (not as part of this PR).

emillon commented on src/asn1.ml:
Cool to see that feature used. Is that the compiler itself that checks `[@tailcall]` or a separate tool? Do you know when it was introduced?

emillon commented on src/asn1.ml:
You might want to extract that length parsing part into a function.

emillon commented on src/asn1.ml:
You can use `Cstruct.BE.get_uint16` here.

emillon commented on src/asn1.ml:
Instead of using a "dummy" OID, what do you think about making that type an option?

emillon commented on src/asn1.ml:
These entries, in a valid key, are always all present sorted by tag number. So you can do something like:

``` ocaml
match symbols with
| [ `Modulus modulus
  ; `Coefficient_a coefficient_a
  ; ...
  ] -> Result.Ok { modulus ; coefficient a ; ... }
| _ -> Result.Error "Parse error: some elements are missing or are not correctly sorted"
```

```
```

emillon commented on src/asn1.ml:
You can remove this case by replacing the `List.find` call above by a function that returns an `oid option` instead of a `[>`Oid of oid]`. It's not in the standard library but you can copy for example [https://github.com/cryptosense/ocamllint/blob/39b1c8c4c4bf95db9d28fdd1ecf12b7d4d2650c4/rules.ml#L209-L216](this one).

emillon commented on src/asn1.mli:
```diff
@@ -50,6 +50,46 @@ sig
   end
 end
 
+module RSA_CVC :
```
Why is this in `Asn1` and not in `Cvc`?

chrismamo1 replied:
Moved everything to the `Key_parsers.Cvc` module.

emillon commented on src/asn1.mli:
```diff
@@ -238,6 +278,14 @@ sig
   val dh_grammar : DH.Params.t Asn.t
 end
 
+module CVC :
```
This sort of API is only useful if you don't know in advance what kind of key you'll parse (ex: in certificates). Here, we'll know out of band what is the kind of key (RSA or ECDSA) from another source, so this is not necessary (that means that the `UNKNOWN` handling is unnecessary).

In other words, we don't have to detect, we know when to call `RSA.parse` or `ECDSA.parse`.

chrismamo1 replied:
Is there a problem with the way it's currently structured, though? Because extracting separate `RSA.parse` and `ECDSA.parse` functions would be a bit tricky given the structure of the parser.

emillon replied:
I do not think that anything will need to call a `CVC.decode` with this signature. So you can create functions with signature `decode : Cstruct.t -> (RSA_CVC.Public.t, string) Result.result` (resp. `ECDSA_CVC.Public.t`) and only expose those.
- In a first step, those can be implemented in terms of the current `decode` function
- In a second step, that should be an opportunity to simplify your code since you do not need to perform detection (that can be done in a later PR).

emillon commented on tests/test_cvc.ml:
```diff
@@ -0,0 +1,109 @@
+  let expected_public =
+    let n =
+      Z.of_string
+        "0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\
```
Using this modulus you won't be able to detect string reversal (e.g. the test suite will pass if you remove the string reversal part). You can use, in order of preference:

-`00 01 02 03...7f 80`
-a random sequence of bytes
-an actual modulus generated for example by openssl. I can provide one if you'd like.

chrismamo1 replied:
Changed to sequential byte sequences.

emillon commented on tests/test_cvc.ml:
```diff
@@ -0,0 +1,109 @@
+          assert_equal ~ctxt ~cmp ~printer ~msg:"e" expected.e real.e
+      | `ECDSA _
+      | `UNKNOWN ->
+          assert_failure ""
```
In that case (you control the data + it's the test suite), you can use `assert false`.

chrismamo1 replied:
Changed, should be reflected in the next push.

emillon commented on tests/test_cvc.ml:
```diff
@@ -0,0 +1,109 @@
+      Z.of_string
+        "0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF"
+    in
+    let coefficient_a =
```
Here you can use actual EC parameters, for example the `secp256r1` ones:
- `modulus = coefficient_a = 0xffffffff00000001000000000000000000000000fffffffffffffffffffffffc`
- `coefficient_b = 0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b`
- `base_point_g = 0x46b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c2964fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5`
- `base_point_r_order = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551`
- `cofactor_f = 1`

And you can use a random value for `cofactor_f` for the same reasons as RSA keys.

chrismamo1 replied:
Changed, should be in the next push (assuming you meant that I should use a random/ascending value for `public_point_y` rather than `cofactor_f`).

emillon replied:
That's what I meant. Thanks!

emillon commented on tests/test_cvc.ml:
```diff
@@ -0,0 +1,109 @@
+    let real = decode cvc in
+    let open Public in
+    Test_util.assert_ok real @@ function
+      | `ECDSA real ->
```
Here you can replace the whole block with `assert_equal ~ctxt ~printer:[%show: ECDSA.t] ~cmp:[%eq: ECDSA.t] expected real` (by adding the `[@@deriving]` declaration where necessary).

emillon replied:
You can even remove the `assert_failure` below by doing the `assert_equal` on the variant type, ie `let expected = ECDSA {modulus; ...}`.

jcm-cryptosense committed:
use option type to handle possible nonexistent OID's

chrismamo1 commented on src/asn1.ml:
Changed, should be in the next push.

chrismamo1 commented on src/asn1.ml:
Thanks, I wasn't sure if that would always be the case so I ran with the (ugly) existing solution. Will change.

jcm-cryptosense committed:
clean up symbol-aggregating code

jcm-cryptosense committed:
clean up option type in OID handling

jcm-cryptosense committed:
extract CVC parser modules from Key_parsers.Asn1 to Key_parsers.Cvc

remove unused and misleading Cvc.RSA functions

change test key parameters from uniform bit strings to sequential bit strings

changed assert_failure to assert false in CVC tests

changed ECDSA CVC test parameters to a named curve

jcm-cryptosense committed:
improve the cvc test library's CVC suite

chrismamo1 commented on src/asn1.ml:
It was added to the language around version `4.02.1` or `4.02.2`. I assume it's handled by the compiler, but I'm not completely sure.

chrismamo1 commented on src/asn1.ml:
Do you mean a function such as this?

``` ocaml
parse_len ?(i = 0) cs =
  let code = Cstruct.get_uint8 cs i in
  if code < 0x80
  then begin
    (* ... *)
  end
```

I'm not sure if that would end up making things much clearer, as I don't think it's used anywhere else, and leaving it inline with the parser allows the eventual extension to arbitrarily long length specifiers.

chrismamo1 commented on src/asn1.ml:
Fixed, should be in next push.

jcm-cryptosense committed:
use Cstruct.BE.get_uint16

emillon commented on src/asn1.ml:
Yes, that what I had in mind. I don't think extracting functions is only for code reuse, but that makes it also limit the size of this (long) function.

To extract it you will need to return both the parsed length and the number of bytes read from the cstruct.

Note that you already need to handle variable-length length specifiers since you need to support specifiers with length 1 and 2 already.

emillon commented on src/cvc.ml:
```diff
@@ -0,0 +1,295 @@
+  | Length
+  | Value of int
+
+let cvc_object_types =
```
There are some unused values here (those who relates to certificates and not keys).

chrismamo1 replied:
Is it particularly harmful to keep them around in case we'd like to add them later? I think it's also nice to avoid jumps in the numbering.

emillon replied:
It is not harmful of course, but parsing certificates is beyond the scope of this library, so they will probably not be added. Their definitions can be pulled from the documentation if required.

chrismamo1 replied:
Fixed.

emillon commented on src/cvc.ml:
There are some leftover debugging printfs here.

jcm-cryptosense committed:
remove leftover debugging printf's

chrismamo1 commented on src/cvc.ml:
Removed.

jcm-cryptosense committed:
remove unused CVC object tags

emillon commented on src/cvc.ml:
Why is that case different?

emillon commented on src/cvc.ml:
What do those `Bytes bytes` entries correspond to?

emillon commented on src/cvc.ml:
What lies behind this `_` case? Can you write it in extension?

emillon replied:
I mean replacing `_` by the list of cases that should be covered by it like:

``` ocaml
(`Something | `SomethingElse) ->
```

chrismamo1 replied:
Changed, turns out that the few cases before the wildcard will catch everything we care about, but the compiler wasn't giving an "unused match case" warning because of the behaviour of inferred polymorphic variants with the wildcard.

emillon replied:
Great, that is was I was hoping!

chrismamo1 commented on src/cvc.ml:
Because `0x82` codes for the exponent in RSA keys and a coefficient in ECDSA.

chrismamo1 commented on src/cvc.ml:
The `V` component of the TLV entry, i.e. `Type` corresponds to a Type token and `Length _` corresponds to a length token. I suppose that for extra safety I could capture the value wrapped by `Length` and check to make sure that `bytes` is actually that long. Would that be worth the extra lines? I think that the parser might already impose enough restrictions up to this point that `bytes` should always have the right length.

chrismamo1 commented on src/cvc.ml:
I'm not quite sure what you mean by writing it "in extension".

emillon commented on src/cvc.mli:
```diff
@@ -0,0 +1,33 @@
+  end
+end
+
+type t =
```
Can you add a `decode` function to the `RSA` and `ECDSA` modules (that constrains the OID) and not expose this one? Thanks!

chrismamo1 replied:
I added `decode` functions to `RSA.Public` and `ECDSA.Public` since that's where the types are defined. I'm not sure if we'll ever have _non_-public CVC keys to parse, though. Would it make sense to move the type+decoder functions directly into `RSA.blah` and `ECDSA.blah`?

chrismamo1 replied:
Also, the `decode` functions are pretty much overlays of the master `decode` function, checking to see that the return value is correctly wrapped. It would be better to modify the master `decode` function to accept an OID as a parameter and terminate parsing if the OID it finds doesn't match the one given, but that sort of modification would have taken far longer and from what I understand speed isn't the highest priority for this feature at this stage.

emillon replied:
Yeah I meant `RSA.Public.decode` indeed. Let's keep the `Public` part. Please make it so that the public API is:
- `Cvc.RSA.Public.decode: Cstruct.t -> (Cvc.RSA.Public.t, string) Result.result`
- `Cvc.ECDSA.Public.decode: Cstruct.t -> (Cvc.ECDSA.Public.t, string) Result.result`
- the existing `Cvc.t` and `Cvc.decode` are _not_ exposed.

chrismamo1 replied:
Yep, that's as I interpreted it. :+1: Pushing now.

emillon commented:
Once the last outstanding issues are fixed (most importantly, fix the API to not rely detection) I think it's good for a first merge. There are some refactorings we can discuss next.

@NathanReb you're OK with this?

emillon commented:
Great job! :+1: :key:

jcm-cryptosense committed:
match in extension while parsing

jcm-cryptosense committed:
move CVC decoders around to clean up library interface

jcm-cryptosense committed:
fix CI build

NathanReb commented on _tags:
```diff
@@ -10,7 +10,10 @@ true: debug
 <src/*.ml>: for-pack(Key_parsers)
 <src/asn1.*>: package(ppx_deriving.std), package(ppx_deriving_yojson)
 <src/ltpa.*>: package(ppx_deriving.ord), package(ppx_deriving_yojson)
+<src/cvc.*>: package(ppx_deriving.ord), package(ppx_deriving_yojson), package(ppx_deriving.eq), package(ppx_deriving.show)
```
Can't `ord`, `eq` and `show` be simplified to `std` ?

chrismamo1 replied:
`std` includes other plugins as well e.g. `enum`, `iter`.

emillon replied:
That looks like premature optimization, `true: package(ppx_deriving.std)` would work as well + be more maintainable. Let's do this in a separate PR.

chrismamo1 replied:
OK. I'll go ahead and make a branch based on this one then.

NathanReb commented on _tags:
```diff
@@ -10,7 +10,10 @@ true: debug
 <src/*.ml>: for-pack(Key_parsers)
 <src/asn1.*>: package(ppx_deriving.std), package(ppx_deriving_yojson)
 <src/ltpa.*>: package(ppx_deriving.ord), package(ppx_deriving_yojson)
+<src/cvc.*>: package(ppx_deriving.ord), package(ppx_deriving_yojson), package(ppx_deriving.eq), package(ppx_deriving.show)
```
Same applies below !

emillon commented:
OK, merging this, thanks.

chrismamo1 commented:
Actually wait a second to merge, I noticed that the CVC code produces a fair number of very easy to fix compiler warnings, about to submit a fix.

emillon commented:
Too late. Also I noticed you modified the wrong changelog entry, I'll fix this on master.
